# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 4d7b95550368
# Full Hash: 4d7b955503687e85bd040d6d359caebb055a0f23
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-01-09 21:39:42
# Regressor Bug: 1600906
# File Overlap Count: 1
# Description:
#   Bug 1600906 - Avoid copying key buffers. r=dom-workers-and-storage-reviewers,ytausky
#   
#   Depends on D57991
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D57992
# ==============================================================================

diff -r 94ca5d86b3b0 -r 4d7b95550368 dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Mon Jan 06 14:19:39 2020 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Mon Jan 06 14:16:33 2020 +0000
@@ -864,12 +864,12 @@
       return NS_ERROR_FILE_CORRUPTED;
     }
 
-    nsCString keyBuffer(reinterpret_cast<const char*>(blobDataIter),
-                        uint32_t(keyBufferLength));
+    IndexDataValue idv(
+        indexId, unique,
+        Key{nsCString{reinterpret_cast<const char*>(blobDataIter),
+                      uint32_t(keyBufferLength)}});
     blobDataIter += keyBufferLength;
 
-    IndexDataValue idv(indexId, unique, Key(keyBuffer));
-
     // Read sort key buffer length.
     const uint64_t sortKeyBufferLength =
         ReadCompressedNumber(&blobDataIter, blobDataEnd);
@@ -883,11 +883,10 @@
         return NS_ERROR_FILE_CORRUPTED;
       }
 
-      nsCString sortKeyBuffer(reinterpret_cast<const char*>(blobDataIter),
-                              uint32_t(sortKeyBufferLength));
+      idv.mLocaleAwarePosition =
+          Key{nsCString{reinterpret_cast<const char*>(blobDataIter),
+                        uint32_t(sortKeyBufferLength)}};
       blobDataIter += sortKeyBufferLength;
-
-      idv.mLocaleAwarePosition = Key(sortKeyBuffer);
     }
 
     if (NS_WARN_IF(!aIndexValues.InsertElementSorted(idv, fallible))) {
@@ -3588,12 +3587,12 @@
       return NS_ERROR_FILE_CORRUPTED;
     }
 
-    nsCString keyBuffer(reinterpret_cast<const char*>(blobDataIter),
-                        uint32_t(keyBufferLength));
+    IndexDataValue idv(
+        indexId, unique,
+        Key{nsCString{reinterpret_cast<const char*>(blobDataIter),
+                      uint32_t(keyBufferLength)}});
     blobDataIter += keyBufferLength;
 
-    IndexDataValue idv(indexId, unique, Key(keyBuffer));
-
     if (blobDataIter < blobDataEnd) {
       // Read either a sort key buffer length or an index id.
       uint64_t maybeIndexId = ReadCompressedNumber(&blobDataIter, blobDataEnd);